### PR TITLE
A capability to add empty directories to a container image

### PR DESCRIPTION
**Pull request by pawelz**

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

File: container/BUILD [hunk 1]
```diff
@@ -102,6 +102,8 @@ TEST_TARGETS = [
     "build_with_tag",
     "with_passwd",
     "with_group",
+    "with_empty_files",
+    "with_empty_dirs",
```
Comment (mattmoor) [inline container/BUILD]: Can you also add tests to the `image_test.py` in this directory that verify these exist in the resulting image?

Comment (pawelz) [reply container/BUILD]: Hi @mattmoor, I am not sure I understand your request. I added a test that verifies that the image contains exactly expected files/dirs, that they have expected mode, and that files are empty. What am I missing?

Comment (mattmoor) [reply container/BUILD]: @pawelz that's perfect, thanks.  Sorry I missed that the first time!

**Comment by dekkagaijin** [comment]
@bazel-io Test this please

**Comment by pawelz** [comment]
@bazel-io Test this please

**Comment by pawelz** [comment]
(Just trying out, whether I can request testing)

**Comment by dekkagaijin** [comment]
Your hubris was your downfall 😆

**Comment by pawelz** [comment]
Hi, is there anything I can do to make the review of this change easier?

**Comment by mattmoor** [comment]
@bazel-io Test this please

File: container/build_tar.py [hunk 2]
```diff
@@ -282,8 +300,7 @@ def main(unused_argv):
     for file_owner in FLAGS.owner_names:
       (f, owner) = file_owner.split('=', 1)
       (user, group) = owner.split('.', 1)
-      if f[0] == '/':
-        f = f[1:]
+      f.lstrip('/')
```
Comment (mattmoor) [inline container/build_tar.py]: doesn't this return the stripped string, vs. modify it in-place?

File: container/build_tar.py [hunk 3]
```diff
@@ -293,29 +310,31 @@ def main(unused_argv):
...
-      if map_filename in mode_map:
-        mode = mode_map[map_filename]
-      if map_filename in ids_map:
-        ids = ids_map[map_filename]
-      if map_filename in names_map:
-        names = names_map[map_filename]
-      output.add_file(inf, tof, mode, ids, names)
+      filename.lstrip('/')
+      if filename in mode_map:
+        mode = mode_map[filename]
```
Comment (mattmoor) [inline container/build_tar.py]: `mode_map.get(filename, default_mode)` is cleaner.

File: container/build_tar.py (hunk 2 above)
Comment (pawelz) [reply container/build_tar.py]: You're absolutely right. Reverted that commit.

It's a bit concerning it didn't break any tests, but that seems outside of the scope of this PR. I'll look into that later.

File: container/build_tar.py (hunk 3 above)
Comment (pawelz) [reply container/build_tar.py]: done

**Comment by pawelz** [comment]
Thanks for the review. Addressed the comments. PTAL.

**Comment by pawelz** [comment]
Hi, can you take another look at this change? I believe I addressed all previous comments. Thanks!

**Comment by pawelz** [comment]
Perfect, thank you!

**Comment by mattmoor** [comment]
@bazel-io Test this please

**Comment by mattmoor** [comment]
@bazel-io Please test this

**Comment by mattmoor** [comment]
@davidstanke what's up with Bazel CI?

**Comment by mattmoor** [comment]
I'm going to close and reopen this PR to attempt to kick Bazel's new CI setup to see this.

**Comment by mattmoor** [comment]
@philwo @buchgr Can you kick CI for this job?

**Comment by pawelz** [comment]
Let me resolve the conflicts first.

**Comment by buchgr** [comment]
Please rebase this pull request to the latest master and push to the branch again. The CI should just work then. Else, please ping me again.

**Comment by pawelz** [comment]
Looks like it worked. Thanks.

**Comment by raonitimo** [comment]
Heya (cc @mattmoor) ! How can I add an owner to an `empty_dir` using `container_image` or `container_layer` ?  It seems support for defining dir owners (with the names argument) and mode was implemented into the `add_empty_dir` method of `TarFile` . But the `container_image` rule (and `container_layer`) supports only `mode`. Do I need to “write a customized container_image” if I want to add support for `owner` of `empty_dirs`?

**Pull request merged**